File: 4-Clustering/1-K-Means_Clustering/src/kmeans.py
# K Means Clustering

# Import the Libraries
import matplotlib.pyplot as plt
import pandas as pd
import os

from sklearn.cluster import KMeans

# Import the data
BinPath = os.path.abspath(os.path.join(os.path.dirname( __file__ ), '..', 'bin'))
FilePath = BinPath + '\Mall_Customers.csv';
dataset = pd.read_csv(FilePath)

X = dataset.iloc[:, [3,4]].values
###Y = dataset.iloc[:, 4].values

# Using the elbow method to find the opimal number of clusters
WCSS = []
for i in range(1, 11):
	kmeans = KMeans(n_clusters = i, 
					init = 'k-means++', 
					max_iter= 300, 
					n_init= 10, 
					random_state=0)
	kmeans.fit(X)
	WCSS.append(kmeans.inertia_)

# Ploting the Elbow Methods
ElbowMethodPlot = plt.figure(1)
plt.plot(range(1,11), WCSS)
plt.title('The Elbow Method')
plt.xlabel('Number of Clusters')
plt.ylabel('WCSS')
ElbowMethodPlot.show()
ElbowMethodPlot.savefig(BinPath + '\ElbowMethod_PY.png')

# Applying k-means to the mall dataset
kmeans = KMeans(n_clusters=5,
				init='k-means++',
				max_iter=300,
				n_init=10,
				random_state=0)
Y_kmeans = kmeans.fit_predict(X)

# Visualising the clusters
KMeansPlot = plt.figure(2)
plt.scatter(X[Y_kmeans == 0, 0], X[Y_kmeans == 0, 1], s = 100, c = 'red', label = 'Careful')
plt.scatter(X[Y_kmeans == 1, 0], X[Y_kmeans == 1, 1], s = 100, c = 'blue', label = 'Standard')
plt.scatter(X[Y_kmeans == 2, 0], X[Y_kmeans == 2, 1], s = 100, c = 'green', label = 'Target')
plt.scatter(X[Y_kmeans == 3, 0], X[Y_kmeans == 3, 1], s = 100, c = 'cyan', label = 'Careless')
plt.scatter(X[Y_kmeans == 4, 0], X[Y_kmeans == 4, 1], s = 100, c = 'magenta', label = 'Sensible')
plt.title('Cluster of clients')
plt.xlabel('Annual Income (k$)')
plt.ylabel('Spending Score (1-10)')
plt.legend()
plt.show()
KMeansPlot.savefig(BinPath + '\KMeansClustering_PY.png')

File: 4-Clustering/1-K-Means_Clustering/src/kmeans.R
# K-Means Clustering

# Path to the file and Libraries
###Change File Directory
BinPath = paste(getwd(), 'MachineLearning-Portfolio/4-Clustering/1-K-Means_Clustering/bin', sep='/')
FilePath = paste(BinPath, '/Mall_Customers.csv', sep='')
setwd(BinPath)

install.packages('caTools', repos = "http://cran.us.r-project.org", dependencies = TRUE)
install.packages('cluster',repos = "http://cran.us.r-project.org", dependencies = TRUE)

library(caTools, help, pos = 2, lib.loc = NULL)
library(cluster)

# Import dataset
dataset <- read.csv(FilePath)
X <- dataset[4:5]

# Using the elbow method to find the optimal number of clusters
png("ElbowMethod_R.png") 
set.seed(6)
WCSS <- vector()
for(i in 1:10) WCSS[i] <- sum(kmeans(X, i)$withinss)
ElbowMethodPlot <- plot(1:10, 
                        WCSS, 
                        type = "b", 
                        main = paste('Clusters of clients'), 
                        xlab = "Number of clusters", 
                        ylab = "WCSS")
dev.off() 

# Apply K-means to the dataset
set.seed(29)
kmeans <- kmeans(X, 5, iter.max = 300, nstart = 10)

# Visualising the clusters
png("KMeansClustering_R.png") 
KMeansPlot <- clusplot(X,
                       kmeans$cluster,
                       lines = 0,
                       shade = TRUE,
                       color = TRUE,
                       labels = 2,
                       plotchar = FALSE,
                       span = TRUE,
                       main = paste('Clusters of clients'),
                       xlab = "Annual Income",
                       ylab = "Spending score (1-100)")
dev.off() 